Mathlib/Algebra/Group/ForwardDiff.lean

```diff
@@ -0,0 +1,178 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Giulio Caflisch, David Loeffler
+-/
+import Mathlib.Algebra.BigOperators.Pi
+import Mathlib.Algebra.Module.Submodule.LinearMap
+import Mathlib.Data.Nat.Choose.Sum
+import Mathlib.Tactic.Abel
+
+/-!
+# Forward difference operators and Newton series
+
+We define the forward difference operator, sending `f` to the function `x ↦ f (x + 1) - f x`
+(for functions on `ℕ`, or more generally on any additive commutative monoid with `1`, taking values
+in an abelian group). The notation `Δ` is defined for this operator, scoped in namespace `fwdDiff`.
+
+We prove two key formulae about this operator:
+
+* `shift_eq_sum_fwdDiff_iter`: the **Gregory-Newton formula**, expressing `f (y + n)` as a linear
+  combination of forward differences of `f` at `y`, for `n ∈ ℕ`;
+* `fwdDiff_iter_eq_sum_shift`: formula expressing the `n`-th forward difference of `f` at `y` as
+  a linear combination of `f (y + k)` for `0 ≤ k ≤ n`.
+
+We also prove some auxiliary results about iterated forward differences of the function
+`n ↦ n.choose k`.
+-/
+
+open Finset Nat Function
+
+variable {M G : Type*} [AddCommMonoidWithOne M] [AddCommGroup G]
+
+/--
+Forward difference operator, `fwdDiff f n = f (n + 1) - f n`. The notation `Δ` for this operator
+is available in the `fwdDiff` namespace.
+-/
+def fwdDiff (f : M → G) : M → G := fun n ↦ f (n + 1) - f n
+
+@[inherit_doc] scoped[fwdDiff] notation "Δ" => fwdDiff
+
+open fwdDiff
+
+@[simp] lemma fwdDiff_add (f g : M → G) : Δ (f + g) = Δ f + Δ g := add_sub_add_comm ..
+
+@[simp] lemma fwdDiff_const (g : G) : Δ (fun _ ↦ g : M → G) = fun _ ↦ 0 := funext fun _ ↦ sub_self g
+
+section smul
+
+lemma fwdDiff_smul {R : Type} [Ring R] [Module R G] (f : M → R) (g : M → G) :
+    Δ (f • g) = Δ f • g + f • Δ g + Δ f • Δ g := by
+  ext y
+  simp only [fwdDiff, Pi.smul_apply', Pi.add_apply, smul_sub, sub_smul]
+  abel
+
+-- Note `fwdDiff_const_smul` is more general than `fwdDiff_smul` since it allows `R` to be a
+-- semiring, rather than a ring; in particular `R = ℕ` is allowed.
+@[simp] lemma fwdDiff_const_smul {R : Type*} [Semiring R] [Module R G] (r : R) (f : M → G) :
+    Δ (r • f) = r • Δ f :=
+  funext fun _ ↦ (smul_sub ..).symm
+
+@[simp] lemma fwdDiff_smul_const {R : Type} [Ring R] [Module R G] (f : M → R) (g : G) :
+    Δ (fun y ↦ f y • g) = Δ f • fun _ ↦ g := by
+  ext y
+  simp only [fwdDiff, Pi.smul_apply', sub_smul]
+
+end smul
+
+section linear_end
+/-
+NB. This contains versions of the forward-difference operator and the shift operator bundled as
+`ℤ`-linear endomorphisms. These are useful for certain proofs; but they are slightly annoying to
+use, as the source and target types of the maps have to be specified each time, and various
+coercions need to be un-wound when the operators are applied. So we conceal them from the user.
+-/
+
+variable (M G) in
+@[simps]
+private def fwdDiffₗ : Module.End ℤ (M → G) where
```

Comment:
I don't really agree with making this private. If it's useful for proofs within this file, it's quite possible it's useful for proofs outside of it.

Reply:
I'd be happy with that, sure! I've un-privatised (publicized?) them, but put them in a namespace, to avoid cluttering up the root namespace too much.